boxoffice/models/line_item.py

```diff
@@ -96,17 +96,23 @@ def calculate(cls, line_item_dicts, coupons=[]):
         """
         item_line_items = {}
         line_items = []
+        coupon_list = list(set(coupons)) if coupons else []
+        discounter = LineItemDiscounter()
+
         for line_item_dict in line_item_dicts:
             item = Item.query.get(line_item_dict['item_id'])
             if not item_line_items.get(unicode(item.id)):
-                item_line_items[unicode(item.id)] = []
-            item_line_items[unicode(item.id)].append(make_ntuple(item_id=item.id,
-                base_amount=item.current_price().amount))
-        coupon_list = list(set(coupons)) if coupons else []
-        discounter = LineItemDiscounter()
+                item_line_items[unicode(item.id)] = {'line_items': [], 'is_available': item.is_available}
+            item_line_items[unicode(item.id)]['line_items'].append(make_ntuple(item_id=item.id,
+                base_amount=item.current_price().amount if item.is_available else None))
+
         for item_id in item_line_items.keys():
-            item_line_items[item_id] = discounter.get_discounted_line_items(item_line_items[item_id], coupon_list)
-            line_items.extend(item_line_items[item_id])
+            if item_line_items[item_id]['is_available']:
+                line_items.extend(discounter.get_discounted_line_items(item_line_items[item_id]['line_items'], coupon_list))
+            else:
+                # item isn't available, no need to calculate discounts
+                line_items.extend(item_line_items[item_id]['line_items'])
```

Comment:
Simplified the flow and added comments - https://github.com/hasgeek/boxoffice/pull/97/files#diff-5e9e8b8df4de8db454de121fd430de77R112